Remove member_id persist & self removal

haorenfsa opened the pull request:
Signed-off-by: shaoyue.chen <shaoyue.chen@zilliz.com>

<!--
 Before you open the request please review the following guidelines and tips to help it be more easily integrated:

 - Describe the scope of your change - i.e. what the change does.
 - Describe any known limitations with your change.
 - Please run any tests or examples that can exercise your modified code.

 Thank you for contributing! We will try to test and integrate the change as soon as we can, but be aware we have many GitHub repositories to manage and can't immediately respond to every request. There is no need to bump or check in on a pull request (it will clutter the discussion of the request).

 Also don't be worried if the request is closed or not integrated sometimes the priorities of Bitnami might not match the priorities of the pull request. Don't fret, the open source community thrives on forks and GitHub makes it easy to keep your changes in a forked repo.
 -->

**Description of the change**

- Remove the `member_id` persist.
- No self removal when pod stop.
- When adding new node, etcdctl will only send requests to active nodes

**Benefits**

- Fix `member_id`, `removal log ` persistence issues.
- In kubernetes, you can kill any number of pods, the etcd cluster can always go back to healthy state. There will be no need for disaster recovery any more
- Fix scaling up failure when some pod is down (while the etcd cluster is healthy)
- Quick scale up: scale up directly to any numbers: eg:  you can scale from 3 nodes to 11 nodes directly instead of scale from 3 to 4 then 4 to 5...

**Possible drawbacks**

- When shrinking the size of etcd-cluster (aka scale down), one have to manually remove the old members from cluster using `etcdctl member remove`

**Applicable issues**

#39 

**Additional information**

see: #39

jotamartos commented:
Hi @haorenfsa,

Thank you for taking the time to create this PR and improve our solution. Without reviewing the code deeply, I can see there is a blocker here that will prevent us from accepting the changes. 

> When shrinking the size of etcd-cluster (aka scale down), one have to manually remove the old members from cluster using etcdctl member remove

Leaving the members information may affect the performance of the deployment and that is not a desired 
configuration. I saw your comment in https://github.com/bitnami/bitnami-docker-etcd/issues/39 but I do not see the benefit of removing this functionality. Could you please provide more details so the team can evaluate properly?  Is there any workaround that can be implemented in the prestop.sh script that makes the self removal process work? 

Thanks

haorenfsa commented:
@jotamartos Hi, thank you for taking the time.

Let's start with the benefit of removing the file writing. 
I hope we have an agreement that the current status of etcd helm chart is quite fragile. Once a pod is down in the etcd-cluster, it can seldom rejoin the cluster again. Our team uses etcd as the backbone of our project. In our practice, it happens 9 times out of 10. Not to say when there're 2 or more pods going down together, the cluster easily die on this.

All the problems leads to the file writings. I think we'd better not bet on the files to be written correctly. We either write more complicated mechanism like what databases are doing to ensure the writing, or we prevent to write at all.

Removing the file writing solves the problem in a comparativly simple way. Like I said ahead, the benefit is that `you can kill any number of pods, the etcd cluster can always go back to healthy state. There will be no need for disaster recovery any more.`

And about removing the prestop.  
We should all agree that etcd is design to persist and ensure consistency. While the `prestop`  removes a replica's data in its shutdown, which is quite the opposite. When pods gets shutdown one after another, it's very likely only one replica is kept at last, we may some suffer data loss. And if the final replicas's disk goes down, we lost all data.
Besides the `prestop` hook has a weak guarantee in kubernetes. And it writes a file, we should prevent it.

About the shrinking, current mechanism didn't guarantee shrinking successfully, either. And I just tested with current version , 3 times and it fails every time. We may do this by emitting an extra `Job` everytime we want to shrink. Besides people don't use this in most cases.

ondrej-smola commented:
First @haorenfsa thank you for getting to the point and your suggestions. @jotamartos my problem is that chart is rather broken (explained above) and IMHO does not hold to the very high production state of other bitnami charts (and we are using many of them). It is the only chart that we had to replace with custom yaml files. My point is that there may need to be some blocker relaxed as to get it to state when it can be relied on in production. I think the changes suggested will get us much closer to the goal. I am more than happy to help testing it.

jotamartos commented:
Hi @haorenfsa  and @ondrej-smola,

Thank you for taking the time to provide this valuable feedback. I created a task on our side to review all this information and to review the PR as well. We will investigate if there is any other improvement we can make to this PR to make the solution more "production ready". I can't provide you with an ETA, but we will notify you with any progress we make. 

Thanks

ondrej-smola commented:
@jotamartos thank you very much, just to add my two cents to automatic member removal: I think most users use etcd as critical piece of infrastructure/distributed system. Most installations will be only 5/3/1 member clusters, mostly 3 or 1. The only reasonable action is to switch between 5 and 3 nodes and even in that case it must be ensured that quorum wont be lost (disruption budgets and others). Switching from 3 to 1 node setup is rare and very specific. I think requiring operator to (semi) manually remove members from etcd cluster is the correct way as we are not speaking about gaining performance by adding etcd nodes - we are gaining availability and this is system reliability decision - not some scaling feature that changes often.

haorenfsa commented:
@jotamartos yeah, we're testing it, too, and together with the path #44. I'll let you know if we found any problem. I'm also glad to help if there's anything else I can do.

javsalgar commented:
Hi,

I've been taking a look and I think it is worth performing these changes. However, as I'm not sure of the potential side effects of breaks in other existing installations, I believe we should have an environment variable that enable or disable the feature:

`ETCD_STORE_MEMBER_ID`: If this variable is true, write the member_id file. In order to not perform a breaking change, default it to true.

In the chart, we can have a value that enables/disables the preStop. 

I think it will be a step forward in the good direction, but for the time being letting users to try and test this feature. If the feedback is positive, we can consider cutting a new major release where we definitely remove all the member_id logic. Thoughts?

haorenfsa commented:
I agree. It's a really good idea to make it an option first.

javsalgar commented:
Awesome! Will you be able to add this to the PR? If not, we continue the work here

haorenfsa commented:
> Awesome! Will you be able to add this to the PR? If not, we continue the work here

Yes, i'm working on it

haorenfsa commented:
@javsalgar Hi, I just updated my code. And I haven't got time to test it yet. I sent you an invitation as collaborator of my forked repo, so that you can edit this patch easily, if you need, check your email.

And here are some description about my updates:

I added 2 environment variables: `ETCD_DISABLE_STORE_MEMBER_ID` & `ETCD_DISABLE_PRESTOP` to controll the feature. And I only updated the `/3.3/debian-10/` folder to avoid the copy for now and for the convenience of review.

I use `ETCD_DISABLE_STORE_MEMBER_ID` instead of `ETCD_STORE_MEMBER_ID` so that it is not adopt by default.
And I add the `ETCD_DISABLE_PRESTOP` in case that users not using helm chart or not using the latest chart version want to use this feature.

When both environment vars are not set, it should work in the same machenism before. I'm not pretty sure whether it would work to only enable one of them.

And in fact my path has a third feature you didn't mention: only send requests to alived nodes in cluster when add new member, check the`add_self_to_cluster()`. I'm not sure whether you want to add it in this patch or open another PR for it.

javsalgar commented:
The changes look good to me, but I would like to test it in our automated system in case there are issues.

ondrej-smola commented:
Will also test it tomorrow

javsalgar commented:
Based on the tests I've done so far, it doesn't seem to break

javsalgar commented on 3.3/debian-10/rootfs/opt/bitnami/scripts/libetcd.sh:
```diff
@@ -685,3 +692,70 @@ etcd_initialize() {
+#########################
+get_member_id() {
+    if ! is_boolean_yes "$ETCD_DISABLE_STORE_MEMBER_ID"; then
+        if ! -s "${ETCD_DATA_DIR}/member_id" then
```
Incorrect syntax

javsalgar commented on 3.3/debian-10/rootfs/opt/bitnami/scripts/libetcd.sh:
```diff
@@ -365,7 +368,7 @@ was_etcd_member_removed() {
         debug "Removal was properly recorded in member_removal.log"
         rm -rf "${ETCD_DATA_DIR:?}/"*
-    elif [[ ! -d "${ETCD_DATA_DIR}/member/snap" ]] && [[ ! -f "$ETCD_DATA_DIR/member_id" ]]; then
+    elif [[ ! -d "${ETCD_DATA_DIR}/member/snap" ]] && [[ is_empty_value "$(get_member_id)" ]]; then
```
Don't use [[ and is_empty_value. Remove the brackets surrounding it

ondrej-smola commented:
Based on tests I made so far it broke in one cases out of 8 ... tested 3 and 5 member clusters. The broken test case is really strange to me. In the end it failed on each member generating distinct clusterId with different memberId - log attached. The bad thing is I am unable to reproduce it again :(.

[etcd-test-2.log](https://github.com/bitnami/bitnami-docker-etcd/files/8965027/etcd-test-2.log)

```
global:
  imageRegistry: harbor.ite.tul.cz
extraEnvVars:
  - name: ETCD_DISABLE_STORE_MEMBER_ID
    value: "true"
  - name: ETCD_DISABLE_PRESTOP
    value: "true"
image:  
  repository: library/etcd
  tag: 3.5.4-no-removal
  pullPolicy: Always
auth:
  rbac:
    rootPassword: root
replicaCount: 3
persistence:
  storageClass: ceph-filesystem
```

haorenfsa commented:
@ondrej-smola Hi, that's due to the issue that #44 is trying to fix, this happens in current version as well. Use this patch and it shall works fine.

ondrej-smola commented:
Rolling updates now just work ... this was the real pain point before :)

ondrej-smola commented:
@haorenfsa thanks - missed that one

ondrej-smola commented:
Repeated tests - LGMT as long as it is released with #44

javsalgar commented:
Thank you so much for the contribution!

arloliu commented:
@haorenfsa 
Enable `ETCD_DISABLE_STORE_MEMBER_ID` will cause this.

```
etcd 17:28:24.87 WARN  ==> You set the environment variable ALLOW_NONE_AUTHENTICATION=yes. For safety reasons, do not use this flag in a production environment.                                                                    
│ etcd 17:28:24.87 INFO  ==> Initializing etcd                                                                                                                                                                                        
│ etcd 17:28:24.87 INFO  ==> Generating etcd config file using env variables                                                                                                                                                          
│ etcd 17:28:24.89 INFO  ==> Detected data from previous deployments                                                                                                                                                                  
│ /opt/bitnami/scripts/libetcd.sh: line 752: ETCD_ACTIVE_ENDPOINTS: unbound variable    
```

haorenfsa commented:
> @haorenfsa Enable `ETCD_DISABLE_STORE_MEMBER_ID` will cause this.
> 
> ```
> etcd 17:28:24.87 WARN  ==> You set the environment variable ALLOW_NONE_AUTHENTICATION=yes. For safety reasons, do not use this flag in a production environment.                                                                    
> │ etcd 17:28:24.87 INFO  ==> Initializing etcd                                                                                                                                                                                        
> │ etcd 17:28:24.87 INFO  ==> Generating etcd config file using env variables                                                                                                                                                          
> │ etcd 17:28:24.89 INFO  ==> Detected data from previous deployments                                                                                                                                                                  
> │ /opt/bitnami/scripts/libetcd.sh: line 752: ETCD_ACTIVE_ENDPOINTS: unbound variable    
> ```

upgrading case, let's discuss it in #46

haorenfsa commented:
And for those who just got here. If you're installing from scratch, I suggest that `ETCD_DISABLE_STORE_MEMBER_ID` & `ETCD_DISABLE_PRESTOP` better be enabled together.
